sweepbatcher/greedy_batch_selection.go

```diff
@@ -0,0 +1,315 @@
+package sweepbatcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/txscript"
+	sweeppkg "github.com/lightninglabs/loop/sweep"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+)
+
+// greedyAddSweep selects a batch for the sweep using the greedy algorithm,
+// which minimizes costs, and adds the sweep to the batch. To accomplish this,
+// it first collects fee details about the sweep being added, about a potential
+// new batch composed of this sweep only, and about all existing batches. Then
+// it passes the data to selectBatch() function, which emulates adding the sweep
+// to each batch and creating new batch for the sweep, and calculates the costs
+// of each alternative. Based on the estimates of selectBatch(), this method
+// adds the sweep to the batch that results in the least overall fee increase,
+// or creates new batch for it. If this method fails for whatever reason, the
+// caller falls back to the simple algorithm (method handleSweep).
+func (b *Batcher) greedyAddSweep(ctx context.Context, sweep *sweep) error {
+	if b.customFeeRate == nil {
+		return errors.New("greedy batch selection algorithm requires " +
```

Comment:
why is custom fee rate required? couldn't you ask the backend for a feerate and run the estimation based on that?

Reply:
This was fixed in the following commits:

 - sweepbatcher: always fill sweep.minFeeRate
 - sweepbatcher: always try greedy batch selection